### PR TITLE
Added support for Concretized Callable for parametric gate

### PR DESCRIPTION
**Pull request by neogyk**
#284 Issue, supporting the Concretized Callable as a parameter for the gate in the circuit and adding the ability to calculate the gradient.

### CODE REVIEW

**Comment by chMoussa** [comment]
Hello @neogyk ,

You'll probably need:
- merge the latest changes from the base branch,
- run `pre-commit run --all-files` for linting before pushing
- add a test then make sure tests pass.
- Remove debugging traces.

File: pyqtorch/differentiation/adjoint.py [hunk 1]
```diff
@@ -58,45 +60,80 @@ def forward(
...
-        ctx.save_for_backward(*param_values)
+        for op in ctx.circuit.flatten()[::-1]:
+            if isinstance(op.param_name, ConcretizedCallable):
+                if op.is_parametric and op.param_name not in ctx.param_names:
+                    torch.set_grad_enabled(True)
+                    values[op.param_name] = op.param_name(values).requires_grad_()
+                    ctx.param_names.append(op.param_name)
+                torch.set_grad_enabled(False)
+
+        ctx.save_for_backward(*values.values())
```
Comment (chMoussa) [inline pyqtorch/differentiation/adjoint.py:79]: Hm I am not sure the forward method has to change. I would say only the backward should. A user could also decide to not differentiate wrt `ConcretizedCallable`, so one should check in the backward if requires_grad is True.

File: pyqtorch/differentiation/adjoint.py [hunk 2]
```diff
@@ -58,45 +60,80 @@ def forward(
...
-                            op.jacobian(values, ctx.embedding),
-                            op.qubit_support,
-                        )
-                        grad = grad_out * 2 * inner_prod(ctx.projected_state, mu).real
-                    if grads_dict[param_name] is not None:
-                        grads_dict[param_name] += grad
-                    else:
-                        grads_dict[param_name] = grad
 
+                    def calculate_grad(argname):
```
Comment (chMoussa) [inline pyqtorch/differentiation/adjoint.py:105]: function needs a typing

File: pyqtorch/differentiation/adjoint.py [hunk 3]
```diff
@@ -58,45 +60,80 @@ def forward(
...
+
                 ctx.out_state = apply_operator(
                     ctx.out_state, op.dagger(values, ctx.embedding), op.qubit_support
                 )
                 if (
                     isinstance(op, (Parametric, HamiltonianEvolution))
                     and op.is_parametric
                 ):
+                    if isinstance(op.param_name, ConcretizedCallable):
+                        grads_dict[op.param_name] = None
```
Comment (chMoussa) [inline pyqtorch/differentiation/adjoint.py:101]: so not sure I follow but why doing so? Feel like again one should just check if the parameters in ConcretizedCallable are among the `grads_dict` keys.

File: pyqtorch/differentiation/adjoint.py [hunk 4]
```diff
@@ -58,45 +60,80 @@ def forward(
...
+                        for argname in op.param_name.independent_args:
+                            grads_dict[argname] = deepcopy(grad)
+                            if grads_dict[argname] is not None:
+                                grads_dict[argname] *= deepcopy(
+                                    torch.autograd.grad(
+                                        values[op.param_name],
+                                        values[argname],
+                                        retain_graph=True,
+                                    )[0]
+                                )
```
Comment (chMoussa) [inline pyqtorch/differentiation/adjoint.py:134]: Could you explain what goes on here? Especially why multiplying deepcopies of grads?

**Review by chMoussa — COMMENTED** [review]
Hello @neogyk , I have a few questions. Note that tests and docs fail in CI. If you use hatch, you can test via:
`hatch -e tests run test
``hatch run docs:serve`

**Comment by chMoussa** [comment]
Hello @neogyk , this is the last day to finish it. Will you be able manage?
Don't forget to pre-commit, fix tests and make sure the doc can be generated without error.

**Comment by chMoussa** [comment]
> @chMoussa, could you please review the changes?

@neogyk could you run precommit?

**Review by chMoussa — APPROVED** [review]
@neogyk Since time is limited and this solution seems to work, I will approve it. I will optimize it further in a later PR. Thanks for your contribution!
